File: 2022/phase2_dress_rehearsal/byuvrx/src/frostlab-inekf-54148f8ac0c7/examples/VictoriaPark.cpp
#include <Eigen/Core>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <iterator>
#include <tuple>
#include <time.h>
#include <unistd.h>

#include <InEKF/Core>
#include <InEKF/SE2Models>
#include <tqdm.h>
#include <matplot/matplot.h>
namespace plt = matplot;

typedef InEKF::SE2<Eigen::Dynamic> SE2_D;

typedef std::pair<int, double> sort_pair;
Eigen::VectorXi solveCostMatrix(Eigen::MatrixXd M){
    int n_mm = M.rows();
    Eigen::VectorXi result(n_mm);

    for(int i=0;i<n_mm;i++){
        Eigen::MatrixXd::Index minMM, minLM;
        M.minCoeff(&minMM, &minLM);
        result[minMM] = minLM;
        M.col(minLM).setConstant(1e8);
        M.row(minMM).setConstant(1e8);
    }
    return result;
}

Eigen::VectorXi dataAssocation(SE2_D state, Eigen::VectorXd zs, InEKF::LandmarkSensor ls){
    int n_lm = state().cols() - 2 - 1;
    int n_mm = zs.size() / 2;

    if(n_lm == 0){
        return Eigen::VectorXi::Ones(n_mm)*-1;
    }

    // hardcode in chi2 values
    double alpha = 5.991464547107979;
    double beta  = 9.21034037197618;
    Eigen::MatrixXd M = Eigen::MatrixXd::Constant(n_mm, n_lm+n_mm, alpha);
    // Make cost matrix
    for(int i=0;i<n_lm;i++){
        ls.sawLandmark(i, state);
        for(int j=0;j<n_mm;j++){
            M(j,i) = ls.calcMahDist(zs.segment<2>(2*j), state);
        }
    }

    Eigen::VectorXi assoc = solveCostMatrix(M);
    
    for(int i=0;i<n_mm;i++){
        // Check if it wants a new landmark
        if(assoc[i] >= n_lm){
            assoc[i] = -1;
            // Also check if there was any other landmarks semi-close to it
            // if so, don't make a new one
            for(int j=0;j<n_lm;j++){
                if(M(i,j) < beta){
                    assoc[i] = -2;
                }
            }
        }
    }
    return assoc;

}

void addLandmark(Eigen::Vector2d z, SE2_D& state){
    double x = state[0][0];
    double y = state[0][1];
    double phi = atan2(state()(1,0), state()(0,0));
    double r = z(0);
    double b = z(1);

    double xl = x + r*cos(b+phi);
    double yl = y + r*sin(b+phi);
    Eigen::Vector2d lm;
    lm << xl, yl;

    state.addCol(lm, Eigen::Matrix2d::Identity()*10000);
}

InEKF::SE2<> makeOdometry(Eigen::Vector2d u, double dt){
    double Ve = u(0);
    double alpha = u(1);

    double a = 3.78;
    double b = 0.50;
    double L = 2.83;
    double H = 0.76;

    double Vc = Ve / (1 - tan(alpha)*H/L);
    double motion_t = Vc/L*tan(alpha);
    double motion_x = Vc - Vc/L*tan(alpha)*b;
    double motion_y = Vc/L*tan(alpha)*a;

    return InEKF::SE2<>(dt*motion_t, dt*motion_x, dt*motion_y);
}

constexpr double pi() { return std::atan(1)*4; }

void loadEvents(std::vector<std::tuple<std::string,double,Eigen::VectorXd>> &events, std::string filename, std::string type){
    // Load file
    std::ifstream file(filename);

    // iterate through each line
    std::string temp;
    while(getline(file, temp)){
        // turn each line into a vector as file
        std::istringstream is( temp );
        std::vector<double> line( ( std::istream_iterator<double>( is ) ), ( std::istream_iterator<double>() ) );
        
        // turn each vector into an eigen vector
        auto n = line.size();
        Eigen::Map<Eigen::VectorXd> data(&line[1], n-1);

        // put into large array of data
        events.push_back( std::make_tuple(type, line[0], data) );
    }
}

int main() {
    /***** SETUP INITIAL STATE *****/
    Eigen::Matrix3d sig = Eigen::Matrix3d::Identity()*.1;
    Eigen::VectorXd init(3);
    init << 45*pi()/180, 0, 0;
    SE2_D x0(init, sig);

    /***** SETUP InEKF *****/
    InEKF::GPSSensor gps(3);
    InEKF::LandmarkSensor laser(0.5, 0.5*pi()/180);

    InEKF::InEKF<InEKF::OdometryProcessDynamic> iekf(x0, InEKF::RIGHT);
    iekf.addMeasureModel("GPS", &gps);
    iekf.addMeasureModel("Laser", &laser);
    Eigen::Vector3d Q;
    Q << 0.5*pi()/180, 0.05, 0.05;
    iekf.pModel->setQ(Q);

    /***** LOAD IN DATA *****/
    std::vector<std::tuple<std::string,double,Eigen::VectorXd>> events;
    loadEvents(events, "../data/victoria_park_ascii/DRS.txt", "odo");
    loadEvents(events, "../data/victoria_park_ascii/GPS.txt", "gps");
    loadEvents(events, "../data/victoria_park_ascii/LASER_TREE.txt", "laser");
    sort(events.begin(), events.end(), [](const auto& lhs, const auto& rhs){
        return std::get<1>(lhs) < std::get<1>(rhs);
    });
    double dt = 0;
    double last_t = 0;
    SE2_D s;

    /***** SETUP PLOT *****/
    auto f = plt::figure(false);
    auto ax = plt::gca();

    std::vector<double> empty = {0};
    std::vector<double> s_x = {0}, s_y = {0};
    std::vector<double> gps_x = {0}, gps_y = {0};

    auto lm_pts  = ax->scatter({-1,1}, {-1,1});
    plt::hold(true);
    auto traj = ax->plot(empty, empty);
    auto gps_pts = ax->scatter(empty, empty);
    plt::legend({lm_pts, traj, gps_pts}, {"Landmarks", "Vehicle", "GPS"});
    f->draw();

    /***** ITERATE THROUGH DATA *****/
    tqdm bar;
    int n=0, N=events.size();
    time_t last_plot = time(0);
    for(auto const& [e, t, data] : events){
        // Odometry
        if(e == "odo"){
            dt = t - last_t;
            last_t = t;

            InEKF::SE2 u = makeOdometry(data, dt);
            s = iekf.Predict(u, dt);
            s_x.push_back( s[0][0] );
            s_y.push_back( s[0][1] );
        }

        // GPS
        if(e == "gps"){
            s = iekf.Update(data, "GPS");
            gps_x.push_back(data[0]);
            gps_y.push_back(data[1]);
            s_x.back() = s[0][0];
            s_y.back() = s[0][1];
        }

        // Laser Measurement
        if(e == "laser"){
            // associate landmarks
            assert(data.size() % 2 == 0);
            int n_mm = data.size() / 2;
            Eigen::VectorXi assoc = dataAssocation(iekf.state_, data, laser);
            // iterate through them
            for(int i=0;i<n_mm;i++){
                if(assoc[i] == -1){
                    addLandmark(data.segment<2>(i*2), iekf.state_);
                    laser.sawLandmark(iekf.state_().cols()-2-1-1, iekf.state_);
                    iekf.Update(data.segment<2>(i*2),"Laser");
                }
                else if(assoc[i] != -2){
                    laser.sawLandmark(assoc[i], iekf.state_);
                    iekf.Update(data.segment<2>(i*2), "Laser");
                }

            }
        }

        // plot once a second
        if(difftime(time(0), last_plot) > 1){
            gps_pts->x_data(gps_x);
            gps_pts->y_data(gps_y);

            traj->x_data(s_x);
            traj->y_data(s_y);

            int n_lm = iekf.state_().cols()-2-1;
            Eigen::VectorXd lm_x = iekf.state_().block(0,3,1,n_lm-1).transpose();
            Eigen::VectorXd lm_y = iekf.state_().block(1,3,1,n_lm-1).transpose();
            lm_pts->x_data(std::vector<double>(lm_x.data(),n_lm+lm_x.data()));
            lm_pts->y_data(std::vector<double>(lm_y.data(),n_lm+lm_y.data()));
            last_plot = time(0);
        }

        bar.progress(n, N);
        n += 1;
        sleep(1e-8);
    }
    bar.finish();
    plt::show();

   return 0;
}

File: 2022/phase2_dress_rehearsal/byuvrx/src/frostlab-inekf-54148f8ac0c7/python/examples/VictoriaPark.py
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm
import matplotlib.markers as mmarkers
from scipy.stats.distributions import chi2
from time import time

np.set_printoptions(suppress=True, linewidth=300)

from inekf import OdometryProcessDynamic, LandmarkSensor, GPSSensor
from inekf import SE2, InEKF, ERROR

def data_association(state, zs, laser_mm):
    n_lm = state.State.shape[0] - 2 - 1
    n_mm = len(zs)

    # if we don't have anything, say they're all new!
    if n_lm == 0:
        return [-1]*n_mm

    alpha = chi2.ppf(0.95, 2)
    beta  = chi2.ppf(0.99, 2)
    A = np.ones((n_mm, n_mm))*alpha
    M = np.zeros((n_mm, n_lm))
    for i in range(0,n_lm):
        laser_mm.sawLandmark(i, state)
        for j, z in enumerate(zs):
            M[j,i] = laser_mm.calcMahDist(z, state)

    M_new = np.hstack((M,A))
    pairs = solve_cost_matrix_heuristic(M_new)
    pairs.sort()

    assoc = [x[1] for x in pairs]

    for i in range(len(assoc)):
        # Check if it wants a new landmark
        if assoc[i] >= n_lm:
            assoc[i] = -1
            # Also check if there was any other landmarks semi-close to it
            # if so, don't make a new one
            for j in range(M.shape[1]):
                if M[i, j] < beta:
                    assoc[i] = -2
                    break

    return assoc

def solve_cost_matrix_heuristic(M):
    n_msmts = M.shape[0]
    result = []

    ordering = np.argsort(M.min(axis=1))

    for msmt in ordering:
        match = np.argmin(M[msmt,:])
        M[:, match] = 1e8
        result.append((msmt, match))

    return result

def addLandmark(z, state):
    x, y = state[0]
    phi = np.arctan2(state.State[1,0],state.State[0,0])
    r, b = z

    xl = x + r*np.cos(b+phi)
    yl = y + r*np.sin(b+phi)

    state.addCol([xl, yl], np.eye(2)*1000)

def x(l):
    return np.array([i[0][0] for i in l])

def y(l):
    return np.array([i[0][1] for i in l])

def makeOdometry(u, dt):
    a = 3.78
    b = 0.50
    L = 2.83
    H = 0.76

    Ve, alpha = u
    Vc = Ve / (1 - np.tan(alpha)*H/L)

    motion = np.zeros(3)
    motion[0] = Vc/L*np.tan(alpha)
    motion[1] = Vc - Vc/L*np.tan(alpha)*b
    motion[2] = Vc/L*np.tan(alpha)*a

    return SE2(dt*motion)

#### SETUP INITIAL STATE
x0 = [45*np.pi/180, 0, 0]
sig = np.diag([.1,.1,.1])
x0 = SE2["D"](x0, sig)

#### SETUP InEKF
gps = GPSSensor(3)
laser = LandmarkSensor(0.5, 0.5*np.pi/180)

iekf = InEKF[OdometryProcessDynamic](x0, ERROR.RIGHT)
iekf.addMeasureModel("GPS", gps)
iekf.addMeasureModel("Laser", laser)
Q = np.diag([0.5*np.pi/180, 0.05, 0.05])
iekf.pModel.setQ(Q)

#### LOAD IN DATA
def read_data(filename):
    with open(filename, 'r') as f:
        data = [[float(x) for x in line.strip().split('\t')] for line in f.readlines()]
    return data

odo_data = np.array(read_data("../../data/victoria_park_ascii/DRS.txt"))
gps_data = np.array(read_data("../../data/victoria_park_ascii/GPS.txt"))
laser_data = read_data("../../data/victoria_park_ascii/LASER_TREE.txt")

events =      [('gps',   x[0], x[1:]) for x in gps_data]
events.extend([('odo',   x[0], x[1:]) for x in odo_data])
events.extend([('laser', x[0], np.array(x[1:]).reshape(-1,2)) for x in laser_data])
events = sorted(events, key=lambda x: x[1])

#### GET PLOT READY
fig, ax = plt.subplots(figsize=(8,6))

traj, = ax.plot([])
veh = ax.scatter([],[], label="Vehicle", marker=">")
gps_pts = ax.scatter([], [], label="GPS", c='g', s=3)
lm_pts  = ax.scatter([], [], label="Landmarks", c='r', s=3)
plt.legend()
plt.show(block=False)

#### ITERATE
states = [x0]
gps_data = []
last_t = 0
last_plot = 0
for i, (e, t, data) in tqdm(enumerate(events), total=len(events)):
    # If it's Odometry
    e, t, data = events[i]
    if e == 'odo':
        dt = t - last_t
        last_t = t

        u = makeOdometry(data, dt)
        s = iekf.Predict(u, dt)
        states.append(s)

    # GPS Measurement
    if e == 'gps':
        s = iekf.Update(data, "GPS")
        states[-1] = s
        gps_data.append(data)

    # Laser Measurement
    if e == 'laser':
        # identify landmarks
        assoc = data_association(iekf.state, data, laser)
        # iterate through them (note data here is still r/b)
        for idx, data in zip(assoc, data):
            if idx == -1:
                addLandmark(data, iekf.state)
                laser.sawLandmark(iekf.state.State.shape[0]-2-1-1, iekf.state)
                iekf.Update(data, "Laser")
            elif idx == -2:
                continue
            else:
                laser.sawLandmark(idx, iekf.state)
                iekf.Update(data, "Laser")

    if time() - last_plot > 1 or i == len(events)-1:
        # plot car
        phi = np.arctan2(states[-1].State[1,0],states[-1].State[0,0])
        m = mmarkers.MarkerStyle(">")
        m._transform = m.get_transform().rotate_deg(phi*180/np.pi)
        veh.set_offsets(states[-1][0])
        veh.set_paths((m.get_path().transformed(m.get_transform()),))
        
        # plot trajectory
        xn, yn = x(states), y(states)
        traj.set_data(xn, yn)

        # plot gps data
        if len(gps_data) != 0:
            curr = gps_pts.get_offsets()
            curr = np.vstack((curr, gps_data))
            gps_pts.set_offsets(curr)
            gps_data = []

        # plot landmark data
        n_lm = iekf.state.State.shape[0] - 2 - 1
        if n_lm != 0:
            lm = np.array([iekf.state[i] for i in range(1,n_lm+1)])
            lm_pts.set_offsets(lm)
        else:
            lm = np.zeros((1,2))

        # set limits
        min_x = min(min(xn), min(lm[:,0]))
        max_x = max(max(xn), max(lm[:,0]))
        min_y = min(min(yn), min(lm[:,1]))
        max_y = max(max(yn), max(lm[:,1]))

        ax.set_xlim(min_x*1.1-1, max_x*1.1+1)
        ax.set_ylim(min_y*1.1-1, max_y*1.1+1)
        plt.draw()
        plt.pause(.001)
        last_plot = time()

plt.show()